P0009/P0009.bs

```diff
@@ -1800,36 +1826,50 @@ Table �: Accessor requirements
 <br/>
 <b>26.7.�.2 Class `accessor_basic` [mdspan.accessor.basic]</b>
 
-1. `accessor_basic` meets the requirements of accessor.
-2. `accessor_basic` gives an accessor that has semantics equivalent to dereferencing a pointer to an array of values.
-3. If `T` is not an object type or is an array type, the program is ill-formed.
+1. `accessor_basic` meets the requirements of accessor policy.
+2. If `T` is not an object type or is an array type, the program is ill-formed.
 
 ```c++
 namespace std {
 namespace experimental {
 namespace fundamentals_v3 {
 
-template<class T>
+template<class ElementType>
 struct accessor_basic {
-  using value_type = T;
-  using pointer = T*;
-  using reference = T&;
+  using offset_policy = accessor_basic;
+  using element_type  = ElementType;
+  using reference     = ElementType&;
+  using pointer       = ElementType*;
+
+  static constexpr typename offset_policy::pointer
+    offset(pointer p, ptrdiff_t i) noexcept;
 
-  // [mdspan.accessor.basic.ops], `accessor_basic` operations
-  constexpr reference operator()(pointer p, ptrdiff_t i) const noexcept;
+  static constexpr reference deref(pointer p,ptrdiff_t i) noexcept;
+
+  static constexpr pointer decay(pointer p) noexcept;
 };
 
 }}}
 ```
 
-<br/>
-<b>26.7.�.2.1 `accessor_basic` operations [mdspan.accessor.basic.ops]</b>
+```c++
+static constexpr typename offset_policy::pointer
+  offset(pointer p,ptrdiff_t i ) noexcept ;
+```
+
+* *Returns:* `p+i` <br/>
+
+```c++
+static constexpr reference deref(pointer p,ptrdiff_t i) noexcept ;
+```
+
+* *Returns:* `p[i]` <br/>
 
 ```c++
-constexpr reference operator()(pointer p, ptrdiff_t i) const noexcept;
+static constexpr pointer decay(pointer p) noexcept ;
```

Comment:
Why is this being made `static`? This changes the type requirement a lot and limits the design space.

Reply:
@brycelelbach See comment above. I said the same thing, and we're changing it.

Reply:
@hcedwar But no other analogous implementation of a customization point like this one in the standard uses a static method here, even though they could. Consider `std::allocator` for instance, which could easily have been defined to have static methods, but it wasn't. If we're going to go against precedent, we need a pretty darn good reason to do so, and I just don't see it. (If we do go against precedent, we need to add a paragraph in a non-normative note explaining the reasoning behind the inconsistency.)

Reply:
@dhollman Do you believe that the operational semantics should specify non-static member requirement whether actually necessary or not?  Or should an implementation be free to fulfill the actual minimal set of operational semantics that satisfy the required interface?

Reply:
I believe the concept table should, e.g., have an expression of the form `a.offset(p,i)`, which can (of course) be fulfilled by a static or non-static method (or any future forms that should arise that meet this syntactic requirement, such as universal call syntax or non-intrusive operator dot). That seems the most consistent with everything else like this in the standard; is there an argument for something else that I'm missing here? (In fact, I don't actually even know how a concept table would enforce a non-static method; how would you even write that requirement?)

Reply:
So it is OK for the `basic_accessor` to define these as static methods.

Reply:
Yes, it is perfectly allowable. However, the existing analog in the standard for a stateless class that meets a concept like this, `std::accessor`, uses instance methods, so we'd have to have a good reason to go against that precedent. Is there such a reason?

Reply:
Please identify the analog / precedent to which you are referring.   I could not find `std::accessor` in N4762.

Reply:
`std::allocator`, sorry.

Reply:
Analogy for precedent doesn't hold.  An `std::allocator` manages, or provides an interface to manage, an underlying resource so there must be state and thus an object abstraction for that mutable state is appropriate.  In contrast an access policy potentially modifies how memory, the state of which is held in a `AccessPolicy::pointer` object, is accessed.  An access policy object should not have mutable state; it might have immutable state if it implements a non-trivial mapping.
My assertion in the original pull request is that such state should be bundled into the `AccessorPolicy::pointer` object which is called an _accessor_.  However, that position was released through compromise on the API design.

Reply:
> An `std::allocator` manages, or provides an interface to manage, an underlying resource so there must be state and thus an object abstraction for that mutable state is appropriate.

By that argument, all of the methods on `layout_left::mapping` and `layout_right::mapping` should be `static`. For that matter, `layout_stride::mapping` represents immutable state. This line of argument conflates `static` and `const`.

These same arguments about immutability apply to dozens of other standard library implementations of customization points that are all specified to use instance methods: `std::less`, `std::equal_to`, `std::default_delete`, `std::hash`, `std::strong_ordering`, just to name a few. I can't think of a single standard library implementation of a customization point (or one from a proposal in flight) that has this form and uses `static` methods.

I could be convinced otherwise if a closer analogy existed that used `static` methods, but all of the ones I can think of (`accessor_traits`, `pointer_traits`, `high_resolution_clock`) are worse analogies than the ones I've provided IMHO. As you know better than I do (because you've been doing this longer), consistency is of the *utmost* importance in library design, and you have to have a *very* good reason to disregard a strong argument based on consistency and uniformity.

> My assertion in the original pull request is that such state should be bundled into the `AccessorPolicy::pointer` object which is called an accessor.

Our objections to this had nothing to do with state or the name *accessor*; it had to do with the customization points being `operator[]` and conversion operators, for which there is no precedent in the standard.